independent-projects/resteasy-reactive/server/runtime/src/main/java/org/jboss/resteasy/reactive/server/mapping/URITemplate.java

```diff
@@ -124,38 +125,52 @@ public URITemplate(String template, boolean prefixMatch) {
         //coalesce the components
         //once we have a CUSTOM_REGEX everything goes out the window, so we need to turn the remainder of the
         //template into a single CUSTOM_REGEX
+        List<String> groupAggregator = null;
         List<String> nameAggregator = null;
         StringBuilder regexAggregator = null;
         Iterator<TemplateComponent> it = components.iterator();
         while (it.hasNext()) {
             TemplateComponent component = it.next();
-            if (nameAggregator != null) {
+            if (groupAggregator != null) {
                 it.remove();
                 if (component.type == Type.LITERAL) {
                     regexAggregator.append(Pattern.quote(component.literalText));
                 } else if (component.type == Type.DEFAULT_REGEX) {
                     regexAggregator.append("(?<").append(component.name).append(">[^/]+?)");
-                    nameAggregator.add(component.name);
+                    groupAggregator.add(component.name);
                 } else if (component.type == Type.CUSTOM_REGEX) {
                     regexAggregator.append("(?<").append(component.name).append(">").append(component.literalText.trim())
                             .append(")");
-                    nameAggregator.add(component.name);
+                    groupAggregator.add(component.name);
                 }
             } else if (component.type == Type.CUSTOM_REGEX) {
                 it.remove();
                 regexAggregator = new StringBuilder();
+                groupAggregator = new ArrayList<>();
                 nameAggregator = new ArrayList<>();
-                regexAggregator.append("(?<").append(component.name).append(">").append(component.literalText.trim())
+
+                // normalize the component name to make it a valid group name, see java.util.Pattern#groupName
+                // group names must start with a character in the range A-Za-z
+                // and after that only contain characters in the range A-Za-z0-9
+                String normalizedCaptureGroupName = "capture";
+                // ensure uniqueness in cases where normalization would result in the same name being used.
+                normalizedCaptureGroupName += component.name.hashCode();
+                normalizedCaptureGroupName += component.name;
+                normalizedCaptureGroupName = normalizedCaptureGroupName.replaceAll("[^a-zA-Z\\d]", "");
+
+                regexAggregator.append("(?<").append(normalizedCaptureGroupName).append(">")
+                        .append(component.literalText.trim())
                         .append(")");
+                groupAggregator.add(normalizedCaptureGroupName + "");
```

Comment:
Might be stupid as I haven't looked at the whole thing but could we simply use index based groups such as `group1` instead of generating something complex?

Reply:
I think for valid names you can reference them later in the regex, so you definitely need to preserve the valid ones. For invalid ones I think anything unique would be fine.

Reply:
But you use the pattern to reference them? It looks extremely weird.

Reply:
So technically if you have @Path("someName:{someRegex}") on the class, you can then use @Path("otherName:{\\ksomeName}") to match whatever the original regex matched.

I don't think this is very common, but its why you need to actually use the name in the regex rather than just an index.

Reply:
I implemented the requested changes. Now the groupname is by default the component name, and only changed to be `group1` (and so on) if it is invalid.